{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[]},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","execution_count":1,"metadata":{"id":"SsnZ4HD-HpKf","executionInfo":{"status":"ok","timestamp":1670927268158,"user_tz":-60,"elapsed":14,"user":{"displayName":"Vittoria Ossanna","userId":"09382633242611702363"}}},"outputs":[],"source":["import pandas as pd\n","from google.colab import drive "]},{"cell_type":"code","source":["DATA_PATH = \"gdrive/MyDrive/DataMining_project/data/\"\n","PATH_DS_1 = DATA_PATH + \"data_tpm_geo/GSE157846_tpm_matrix_all_samples_v2.csv\"\n","PATH_DS_2 = DATA_PATH + \"data_tpm_geo/GSE126797_TPM_matrix_all_datasets.csv\"\n","PATH_DS_3 = DATA_PATH + \"data_tpm_geo/GSM4838073_TPM_matrix_Breast_n3.csv\"\n","PATH_SNODB = DATA_PATH + \"data_snodb/snoDB_All_V2.0.tsv\"\n","PATH_RIBO = DATA_PATH + \"data_db_annotation/ensembl_ribosomal_proteins.tsv\"\n","\n","COL_NAMES = [\n","    \"gene_id\",           \"gene_name\",         \"gene_biotype\",\n","    \"Testis_1\",          \"Testis_2\",          \"Testis_3\", \n","    \"Skeletal_muscle_1\", \"Skeletal_muscle_2\", \"Skeletal_muscle_3\",\n","    \"Liver_1\",           \"Liver_2\",           \"Liver_3\",\n","    \"Brain_1\",           \"Brain_2\",           \"Brain_3\", \n","    \"Breast_1\",          \"Breast_2\",          \"Breast_3\",\n","    \"Prostate_1\",        \"Prostate_2\",        \"Prostate_3\",\n","    \"Ovary_1\",           \"Ovary_2\",           \"Ovary_3\"\n","]"],"metadata":{"id":"jsTMHN_AASDM","executionInfo":{"status":"ok","timestamp":1670927268159,"user_tz":-60,"elapsed":13,"user":{"displayName":"Vittoria Ossanna","userId":"09382633242611702363"}}},"execution_count":2,"outputs":[]},{"cell_type":"markdown","source":["# Preliminary steps\n"],"metadata":{"id":"8QqM6fds6BOt"}},{"cell_type":"markdown","source":["Allow the script to access the drive in order to load the files. It might require some time. It is assumed that you added the \"DataMining_poject\" folder as a shortcut (simply right click on the folder > \"Add shortcut to drive\")."],"metadata":{"id":"4nAzyXWU3QES"}},{"cell_type":"code","source":["drive.mount('/content/gdrive')"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"YztfI25o8SYd","executionInfo":{"status":"ok","timestamp":1670927295362,"user_tz":-60,"elapsed":27213,"user":{"displayName":"Vittoria Ossanna","userId":"09382633242611702363"}},"outputId":"f002c104-4d67-495a-f831-b3b586c974d0"},"execution_count":3,"outputs":[{"output_type":"stream","name":"stdout","text":["Mounted at /content/gdrive\n"]}]},{"cell_type":"markdown","source":["**Load data sets** which are stored with the respective GSE name. Notice that:\n","- **ds1** contains expression levels for testis, skeletal muscle, liver and brain\n","  (3 samples each accordingly names). Moreover it contains gene names and \n","  gene biotypes.\n","- **ds2** contains expression levels for breast (BRN117 and BRN491), prostate \n","  (NPR026, NPR029, NPR036), ovary (OVN218, OVN377, OVN85) and some cell-line\n","  samples that are unrelated with the experiment of interest\n","- **ds3** contains expression levels for one breast sample\n","\n","The numbering of the samples does NOT match that of the authors (it should not matter, eventually retrieve it if batch effects are found and further investigations are required)\n"],"metadata":{"id":"Bmzvhtj-CZvg"}},{"cell_type":"code","source":["ds1 = pd.read_csv(PATH_DS_1)\n","ds2 = pd.read_csv(PATH_DS_2)\n","ds3 = pd.read_csv(PATH_DS_3)"],"metadata":{"id":"d6jTU9bg6A-t","executionInfo":{"status":"ok","timestamp":1670927298379,"user_tz":-60,"elapsed":3025,"user":{"displayName":"Vittoria Ossanna","userId":"09382633242611702363"}}},"execution_count":4,"outputs":[]},{"cell_type":"markdown","source":["Load data from snoDB"],"metadata":{"id":"8Cb4Mogr9GPm"}},{"cell_type":"code","source":["snodb = pd.read_csv(PATH_SNODB, sep=\"\\t\")"],"metadata":{"id":"eZKAyIRo9Ijo","executionInfo":{"status":"ok","timestamp":1670927299048,"user_tz":-60,"elapsed":679,"user":{"displayName":"Vittoria Ossanna","userId":"09382633242611702363"}}},"execution_count":5,"outputs":[]},{"cell_type":"markdown","source":["# snoRNA: Merge and filter"],"metadata":{"id":"E3d1CriV3FbP"}},{"cell_type":"markdown","source":["**Merge datasets**, remove not-needed columns and rename rows and columns accordingly"],"metadata":{"id":"927ZbjhJD3At"}},{"cell_type":"code","source":["ds_merged = ds1.merge(ds3, on=\"gene_id\", how=\"outer\")\n","ds_merged = ds_merged.merge(ds2, on=\"gene_id\", how=\"outer\")\n","ds_merged.drop(list(ds_merged.filter(regex = 'SKOV')), axis = 1, inplace = True)\n","ds_merged.columns = COL_NAMES\n","ds_merged.set_index(\"gene_id\", inplace = True)\n","\n","ds_merged.to_csv('tpm_matrix.csv')"],"metadata":{"id":"lyrdf4YJDnnJ","executionInfo":{"status":"ok","timestamp":1670927301857,"user_tz":-60,"elapsed":2816,"user":{"displayName":"Vittoria Ossanna","userId":"09382633242611702363"}}},"execution_count":6,"outputs":[]},{"cell_type":"markdown","source":["**Create two subsets**, one containing gene information, one containing the transcript expression levels."],"metadata":{"id":"Y3htE-UOzf0V"}},{"cell_type":"code","source":["ds_info = ds_merged.iloc[:, :2]\n","ds_exp = ds_merged.iloc[:, 2:]"],"metadata":{"id":"jDjvC-uFGmdP","executionInfo":{"status":"ok","timestamp":1670927301859,"user_tz":-60,"elapsed":8,"user":{"displayName":"Vittoria Ossanna","userId":"09382633242611702363"}}},"execution_count":7,"outputs":[]},{"cell_type":"markdown","source":["**Create a function** that, given a set of parameters, filters the dataset and returns: \n","- the filtered dataset\n","- a dictionary containing\n","    - number of genes\n","    - number of snoRNA\n","    - number of host genes \n","    - number of pairings of snoRNA-host genes\n","    - number of host-less snoRNA \n","    - number of snoRNA and host-genes with missing values"],"metadata":{"id":"p8IjxB1S9Ps1"}},{"cell_type":"code","source":["def filter_ds(\n","    ds_to_filter: pd.DataFrame,\n","    db_snorna: pd.DataFrame,\n","    min_exp: float,\n","    min_samples: float\n","):\n","    \n","    \"\"\"\n","    Filter a dataset and return some useful information.\n","\n","    Given a dataframe representing an expression matrix, filter it according to\n","    some given parameters, then return the filtered dataframe and some \n","    useful information (such as number of snoRNAs etc...). \n","\n","    Parameters\n","    ----------\n","    ds_to_filter : pd.DataFrame\n","        Pandas dataframe with ENSEMBL genes as rows and samples as columns. \n","    db_snorna : pd.DataFrame\n","        Pandas dataframe containing snoDB information, with genes as rows and\n","        gene attributes as columns.\n","    min_exp : float\n","        Float representing the number of TPM to consider a gene expressed.\n","    min_samples : float\n","        Float representing the fraction of samples that must have the gene\n","        expressed in order for the gene to be retained.\n","\n","    Returns\n","    -------\n","    ds_filtered : pd.DataFrame\n","        Pandas dataframe corresponding to the starting dataframe filtered using\n","        the provided parameters.\n","    filt_info : dict\n","        Dictionary containing information about the filtering process.\n","    \"\"\"\n","    \n","    # Compute how many samples have gene expression geq than the threshold for \n","    # each gene, than if it is geq than the minimal required number of samples\n","    # keep it, else discard it\n","    req_samples = ds_to_filter.shape[1] * min_samples // 1\n","    gene_counts = (ds_to_filter >= min_exp).sum(axis = 1)\n","    indexer = gene_counts >= req_samples\n","    ds_filtered = ds_to_filter[indexer]\n","    \n","    # Initialize counters to increment\n","    filt_info = {\n","        \"num_genes\": 0,\n","        \"num_snorna\": 0,\n","        \"num_hosts\": 0,\n","        \"num_host_less\": 0,\n","        \"num_paired\": 0,\n","        \"num_with_nans\": 0,\n","    }\n","\n","    genes_filt = ds_filtered.index.values  # List of genes in filtered list\n","    snorna_list = list(db_snorna[\"ensembl_id\"])  # List of snorna in db\n","    host_list = list(db_snorna[\"host_gene_id\"])  # List of hosts in db\n","    \n","    filt_info[\"num_genes\"] = len(genes_filt)  # Final number of genes kept\n","\n","    for gene in genes_filt:\n","        \n","        significant = False  # Is either snorna gene or host\n","        \n","        # If gene is a snorna\n","        if gene in snorna_list:\n","            significant = True\n","            filt_info[\"num_snorna\"] += 1\n","            \n","            # Retrieve host gene\n","            pos = db_snorna[\"ensembl_id\"]  \n","            host_gene = db_snorna.loc[pos == gene, \"host_gene_id\"].item()\n","\n","            if host_gene in genes_filt:  # Host gene is in filtered genes\n","                filt_info[\"num_paired\"] += 1\n","\n","            if type(host_gene) == float:  # If host gene type is NaN \n","                filt_info[\"num_host_less\"] += 1\n","\n","        # If gene is a host gene\n","        if gene in host_list:\n","            significant = True\n","            filt_info[\"num_hosts\"] += 1\n","        \n","        # If gene is either a snorna or a host gene\n","        if significant and ds_filtered.loc[[gene]].isnull().values.any():\n","            filt_info[\"num_with_nans\"] += 1      \n","\n","    return ds_filtered, filt_info\n"],"metadata":{"id":"BtlxI0yQ9PVh","executionInfo":{"status":"ok","timestamp":1670927301860,"user_tz":-60,"elapsed":8,"user":{"displayName":"Vittoria Ossanna","userId":"09382633242611702363"}}},"execution_count":8,"outputs":[]},{"cell_type":"markdown","source":["**Iterate through some pairs of parameters** (min_exp, min_samples) and print in a tabular way the information."],"metadata":{"id":"iria5Eh02dHn"}},{"cell_type":"code","source":["PARAMS = (\n","    (0, 0.0),\n","    (1, 0.1),\n","    (1, 0.2),\n","    (1, 0.3),\n","    (2, 0.2),\n","    (2, 0.5),\n",")\n","\n","infos = list()\n","\n","for params in PARAMS:\n","    _, info = filter_ds(ds_exp, snodb, *params)\n","    infos.append(info)\n","\n","info_ds = pd.DataFrame.from_dict(infos)\n","info_ds.index = [str(par) for par in PARAMS]\n","\n","info_ds"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":417},"id":"h6FNz8D3DdXK","executionInfo":{"status":"ok","timestamp":1670927322383,"user_tz":-60,"elapsed":20529,"user":{"displayName":"Vittoria Ossanna","userId":"09382633242611702363"}},"outputId":"d07b0ea9-2127-457a-cab6-66f32ca4da9f"},"execution_count":9,"outputs":[{"output_type":"execute_result","data":{"text/plain":["          num_genes  num_snorna  num_hosts  num_host_less  num_paired  \\\n","(0, 0.0)      58884         951       1045            371         571   \n","(1, 0.1)      19341         420        913             37         363   \n","(1, 0.2)      13509         396        800             29         339   \n","(1, 0.3)      11690         385        738             25         329   \n","(2, 0.2)       9492         378        622             20         320   \n","(2, 0.5)       5145         345        343             11         252   \n","\n","          num_with_nans  \n","(0, 0.0)              1  \n","(1, 0.1)              0  \n","(1, 0.2)              0  \n","(1, 0.3)              0  \n","(2, 0.2)              0  \n","(2, 0.5)              0  "],"text/html":["\n","  <div id=\"df-ea110cc7-a941-4869-bb83-196bcc13fb6d\">\n","    <div class=\"colab-df-container\">\n","      <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>num_genes</th>\n","      <th>num_snorna</th>\n","      <th>num_hosts</th>\n","      <th>num_host_less</th>\n","      <th>num_paired</th>\n","      <th>num_with_nans</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>(0, 0.0)</th>\n","      <td>58884</td>\n","      <td>951</td>\n","      <td>1045</td>\n","      <td>371</td>\n","      <td>571</td>\n","      <td>1</td>\n","    </tr>\n","    <tr>\n","      <th>(1, 0.1)</th>\n","      <td>19341</td>\n","      <td>420</td>\n","      <td>913</td>\n","      <td>37</td>\n","      <td>363</td>\n","      <td>0</td>\n","    </tr>\n","    <tr>\n","      <th>(1, 0.2)</th>\n","      <td>13509</td>\n","      <td>396</td>\n","      <td>800</td>\n","      <td>29</td>\n","      <td>339</td>\n","      <td>0</td>\n","    </tr>\n","    <tr>\n","      <th>(1, 0.3)</th>\n","      <td>11690</td>\n","      <td>385</td>\n","      <td>738</td>\n","      <td>25</td>\n","      <td>329</td>\n","      <td>0</td>\n","    </tr>\n","    <tr>\n","      <th>(2, 0.2)</th>\n","      <td>9492</td>\n","      <td>378</td>\n","      <td>622</td>\n","      <td>20</td>\n","      <td>320</td>\n","      <td>0</td>\n","    </tr>\n","    <tr>\n","      <th>(2, 0.5)</th>\n","      <td>5145</td>\n","      <td>345</td>\n","      <td>343</td>\n","      <td>11</td>\n","      <td>252</td>\n","      <td>0</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>\n","      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-ea110cc7-a941-4869-bb83-196bcc13fb6d')\"\n","              title=\"Convert this dataframe to an interactive table.\"\n","              style=\"display:none;\">\n","        \n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","       width=\"24px\">\n","    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n","    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n","  </svg>\n","      </button>\n","      \n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      flex-wrap:wrap;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","      <script>\n","        const buttonEl =\n","          document.querySelector('#df-ea110cc7-a941-4869-bb83-196bcc13fb6d button.colab-df-convert');\n","        buttonEl.style.display =\n","          google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","        async function convertToInteractive(key) {\n","          const element = document.querySelector('#df-ea110cc7-a941-4869-bb83-196bcc13fb6d');\n","          const dataTable =\n","            await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                     [key], {});\n","          if (!dataTable) return;\n","\n","          const docLinkHtml = 'Like what you see? Visit the ' +\n","            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","            + ' to learn more about interactive tables.';\n","          element.innerHTML = '';\n","          dataTable['output_type'] = 'display_data';\n","          await google.colab.output.renderOutput(dataTable, element);\n","          const docLink = document.createElement('div');\n","          docLink.innerHTML = docLinkHtml;\n","          element.appendChild(docLink);\n","        }\n","      </script>\n","    </div>\n","  </div>\n","  "]},"metadata":{},"execution_count":9}]},{"cell_type":"markdown","source":["# snoRNA: Export"],"metadata":{"id":"tx2jkUHC17HL"}},{"cell_type":"markdown","source":["**Rerun** using final parameters"],"metadata":{"id":"EmFSJ0mnwYll"}},{"cell_type":"code","source":["final_params = (1, 0.2)\n","filt_mat, filt_info = filter_ds(ds_exp, snodb, *final_params)\n","filt_mat.dropna(inplace = True)\n","filt_mat = filt_mat.round(2)"],"metadata":{"id":"RZdM5GjWcwWO","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1670927324484,"user_tz":-60,"elapsed":2108,"user":{"displayName":"Vittoria Ossanna","userId":"09382633242611702363"}},"outputId":"257f7d5c-20ff-4cc9-9945-067036356c46"},"execution_count":10,"outputs":[{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.8/dist-packages/pandas/util/_decorators.py:311: SettingWithCopyWarning: \n","A value is trying to be set on a copy of a slice from a DataFrame\n","\n","See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n","  return func(*args, **kwargs)\n"]}]},{"cell_type":"markdown","source":["**Save** expression matrix data to csv"],"metadata":{"id":"a9c4dzbiwc2F"}},{"cell_type":"code","source":["filt_mat.to_csv(DATA_PATH + \"data_expansion_input/filtered_dataset.csv\")"],"metadata":{"id":"VOSBxngMu5nh","executionInfo":{"status":"ok","timestamp":1670927325415,"user_tz":-60,"elapsed":937,"user":{"displayName":"Vittoria Ossanna","userId":"09382633242611702363"}}},"execution_count":11,"outputs":[]},{"cell_type":"markdown","source":["Retrieve **list of snoRNA to expand**"],"metadata":{"id":"HyxT_qg-wveU"}},{"cell_type":"code","source":["snorna_list = [\n","    snorna \n","    for snorna in filt_mat.index \n","    if snorna in list(snodb[\"ensembl_id\"])\n","]"],"metadata":{"id":"nf-7ALOBwvGK","executionInfo":{"status":"ok","timestamp":1670927329510,"user_tz":-60,"elapsed":4097,"user":{"displayName":"Vittoria Ossanna","userId":"09382633242611702363"}}},"execution_count":12,"outputs":[]},{"cell_type":"markdown","source":["Save list of snoRNA **in plain txt format**"],"metadata":{"id":"FDgIpwNkGhjN"}},{"cell_type":"code","source":["with open(DATA_PATH + \"data_expansion_input/exp_snorna.txt\", \"w\") as txt_file:\n","    for snorna in snorna_list:\n","        txt_file.write(f\"{snorna}\\n\")"],"metadata":{"id":"ylRv2wkBGg2_","executionInfo":{"status":"ok","timestamp":1670927329510,"user_tz":-60,"elapsed":23,"user":{"displayName":"Vittoria Ossanna","userId":"09382633242611702363"}}},"execution_count":13,"outputs":[]},{"cell_type":"markdown","source":["Save list of **snoRNA in lgn format**"],"metadata":{"id":"RQql92pg0kti"}},{"cell_type":"code","source":["with open(DATA_PATH + \"data_expansion_input/snorna_net.lgn\", \"w\") as lgn_file:\n","    lgn_file.write(\"from,to\\n\")\n","    for snorna in snorna_list:\n","        lgn_file.write(f\"{snorna},{snorna}\\n\")"],"metadata":{"id":"PAiqd3Fv0n9Z","executionInfo":{"status":"ok","timestamp":1670927329512,"user_tz":-60,"elapsed":20,"user":{"displayName":"Vittoria Ossanna","userId":"09382633242611702363"}}},"execution_count":14,"outputs":[]},{"cell_type":"markdown","source":["# ribosomal proteins: Export"],"metadata":{"id":"R5aiSJBj11tS"}},{"cell_type":"markdown","source":["Read list of all ribosomal proteins according to HUGO"],"metadata":{"id":"g5I_qGRG19qx"}},{"cell_type":"code","source":["all_ribo = pd.read_csv(PATH_RIBO, sep=\"\\t\")"],"metadata":{"id":"I8mrz0gU183E","executionInfo":{"status":"ok","timestamp":1670927329514,"user_tz":-60,"elapsed":19,"user":{"displayName":"Vittoria Ossanna","userId":"09382633242611702363"}}},"execution_count":15,"outputs":[]},{"cell_type":"markdown","source":["Filter ensemble ids of non-mitochondrial ribo-proteins"],"metadata":{"id":"rAi56_RX3GkP"}},{"cell_type":"code","source":["non_mito_ribo = all_ribo[~all_ribo['Group name'].str.contains(\n","    \"mitochondrial\", case=False)]"],"metadata":{"id":"vsh2on0t3KI9","executionInfo":{"status":"ok","timestamp":1670927329515,"user_tz":-60,"elapsed":18,"user":{"displayName":"Vittoria Ossanna","userId":"09382633242611702363"}}},"execution_count":16,"outputs":[]},{"cell_type":"markdown","source":["Filter out only ribosomal proteins present in the filtered dataset"],"metadata":{"id":"oj31w3LSOGIY"}},{"cell_type":"code","source":["ribo_list = [\n","    ribo_p\n","    for ribo_p in list(non_mito_ribo[\"Ensembl gene ID\"])\n","    if ribo_p in filt_mat.index \n","]"],"metadata":{"id":"5qF-F-rj4gg9","executionInfo":{"status":"ok","timestamp":1670927329516,"user_tz":-60,"elapsed":17,"user":{"displayName":"Vittoria Ossanna","userId":"09382633242611702363"}}},"execution_count":17,"outputs":[]},{"cell_type":"markdown","source":["Save list of ribosomal proteins to expand in plain txt format"],"metadata":{"id":"Sy1B5j5VOh4K"}},{"cell_type":"code","source":["with open(DATA_PATH + \"data_expansion_input/ribosomal_list.txt\", \"w\") as txt:\n","    for ribo in ribo_list:\n","        txt.write(f\"{ribo}\\n\")"],"metadata":{"id":"-RR9RbUc5r7j","executionInfo":{"status":"ok","timestamp":1670927330134,"user_tz":-60,"elapsed":633,"user":{"displayName":"Vittoria Ossanna","userId":"09382633242611702363"}}},"execution_count":18,"outputs":[]},{"cell_type":"markdown","source":["Save list of **ribosomal proteins in lgn format**"],"metadata":{"id":"cVYUhTjhO5d9"}},{"cell_type":"code","source":["with open(DATA_PATH + \"data_expansion_input/ribosomal_net.lgn\", \"w\") as lgn:\n","    lgn.write(\"from,to\\n\")\n","    for ribo_p in ribo_list:\n","        lgn.write(f\"{ribo_p},{ribo_p}\\n\")"],"metadata":{"id":"BUnckVFwPT0O","executionInfo":{"status":"ok","timestamp":1670927330136,"user_tz":-60,"elapsed":9,"user":{"displayName":"Vittoria Ossanna","userId":"09382633242611702363"}}},"execution_count":19,"outputs":[]},{"cell_type":"markdown","source":["Save list of **ribosomal and snoRNA in a single lgn** file"],"metadata":{"id":"ZzV5juhJRREh"}},{"cell_type":"code","source":["snorna_ribo_list = snorna_list + ribo_list\n","with open(DATA_PATH + \"data_expansion_input/snorna_ribo_net.lgn\", \"w\") as lgn:\n","    lgn.write(\"from,to\\n\")\n","    for sno_ribo in snorna_ribo_list:\n","        lgn.write(f\"{sno_ribo},{sno_ribo}\\n\")"],"metadata":{"id":"iVdoPRzHRQgG","executionInfo":{"status":"ok","timestamp":1670927330514,"user_tz":-60,"elapsed":386,"user":{"displayName":"Vittoria Ossanna","userId":"09382633242611702363"}}},"execution_count":20,"outputs":[]}]}